### ISSUE TITLE
Godot 4: Some methods in Array doesn't preserve typing

### ISSUE THREAD

**Issue by gustavi** [godotengine/godot#56075]
### Godot version

v4.0.dev.20211210.official [21b81998b]

### System information

Windows 10 64bit - Vulkan API 1.2.141 - GeForce GTX 1660 SUPER

### Issue description

When you type an `Array` with `Array[<inner type>]` the `Array` returned by `.slice()` method has the wrong type (it looks like the type is only `Array`).

Tested methods:

- `duplicate()` : not affected (tested with deep=true and deep=false)
- `filter()` : **affected**
- `map()` : **affected**
- `slice()` : **affected**

![image](https://user-images.githubusercontent.com/2736557/146668550-8075436e-ae9e-421e-b889-22290afcb239.png)

### Steps to reproduce

1. Create a variable `a` typed with `Array[something]` (you can use `Vector2i`)
2. Create a function `f(arg: Array[something])`
3. Create a variable `b` from `a.slice(...)`
4. Try to call `f(b)`

### Minimal reproduction project

This following script is enough to reproduce:

```gdscript
extends Node2D

func _filter(x):
    return true

func _map(x):
    return x

func foo(bar: Array[Vector2i]):
    # Affected
    return bar.map(_map)
    # return bar.filter(_filter)
    # return bar.slice(0, len(bar))

    # Not Affected
    # return bar.duplicate()

func _ready():
    var a: Array[Vector2i] = [Vector2i(0, 0), Vector2i(1, 1), Vector2i(2, 2)]
    var b = foo(a)
    var c = foo(b)
```

**Comment by cdemirer** [comment]
I agree about `slice()` and `filter()`, but maybe it's better that `map()` returns `Array`.

**Comment by kleonc** [comment]
Fixed by #67103 (which fixed #67015 which was a duplicate of this issue).
